### PR TITLE
Added possiblity to generate thumbnails/previews for uploaded videos using FFMPEG

### PR DESCRIPTION
**Pull request by stemod**
This branch adds the possibility to generate preview images/thumbnails for uploaded videos.

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 1]
```diff
@@ -111,6 +111,11 @@ class FileVersion implements AuditableInterface
     private $defaultMeta;
 
     /**
+     * @var string
+     */
+    private $properties = '';
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: shouldn't it be `{}`? else getProperties will return null instead of an array.

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 2]
```diff
@@ -642,4 +647,24 @@ public function getStorageType()
     {
         return $this->storageType;
     }
+
+    /**
+     * @return array
+     */
+    public function getProperties()
+    {
+        return json_decode($this->properties, JSON_OBJECT_AS_ARRAY);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: The second parameter is a boolean, just pass `true` there.

**Comment by alexander-schranz** [comment]
looks good for me added an issue for a more dynamic way to include things like this #1659.
maybe @danrot or @wachterjohannes could also have a look at this PR.

File: src/Sulu/Bundle/MediaBundle/Media/Video/FFMPEGToolBox.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Video;
+
+class FFMPEGToolBox implements FFMPEGToolBoxInterface
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/FFMPEGToolBox.php]: I would call this class `VideoUtils` or something like this... And in general, isn't there a ready to use library or bundle to use, instead of implementing stuff like this on your own?

File: src/Sulu/Bundle/MediaBundle/Media/Video/FFMPEGToolBoxInterface.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Video;
+
+interface FFMPEGToolBoxInterface
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/FFMPEGToolBoxInterface.php]: Same as above

**Comment by danrot** [comment]
@alexander-schranz You issue is something we should really do in some way at some time. We should also extract the ghostscript stuff like this.

@stemod From what I've seen you have only adapted the existing tests, but the tests for your new classes are missing. You should probably wait with that until we've decided if we should use an external library, a quick search revealed https://github.com/pulse00/ffmpeg-bundle, which is a bundle integrating the library https://github.com/PHP-FFMpeg/PHP-FFMpeg. Would be much nicer than execute these command on our own spread over different classes.

**Comment by stemod** [comment]
FFMPEG can return different values for durations. (microseconds)

Please let me know if there´s a good solution to improve this test:

1) Sulu\Bundle\MediaBundle\Tests\Functional\Video\VideoUtilsTest::testGetVideoDuration
Video duration does not match
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'00:09:56.45'
+'00:09:56.46'

File: .travis.yml [hunk 5]
```diff
@@ -1,4 +1,4 @@
-sudo: false
+sudo: true
```
Comment (danrot) [inline .travis.yml]: Why is this necessary? That brings a bunch of disadvantages.

**Comment by danrot** [comment]
@stemod How is it possible that the duration of the video differs? If there is really no correct solution to this, some of the assertions allow to pass delta values.

File: .travis.yml (hunk 5 above)
Comment (wachterjohannes) [inline .travis.yml]: if we want to install something in the before_install section you have to set this.

Comment (danrot) [inline .travis.yml]: That's another reason to use https://github.com/PHP-FFMpeg/PHP-FFMpeg, in that case this can easily be mocked, and we don't have to do this. Then the tests can still be ran with the new travis infrastructure.

Comment (wachterjohannes) [inline .travis.yml]: hmm i have to say that this library is such easy to use we should invest that hour to make it work (@stemod i think its not more time)

File: src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php [hunk 6]
```diff
@@ -92,11 +93,15 @@ class FormatManager implements FormatManagerInterface
      */
     private $formats;
 
+    /** @var VideoThumbnailServiceInterface */
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php]: Split this comment over three lines.

File: src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php [hunk 7]
```diff
@@ -528,4 +542,22 @@ public function purge($idMedia, $fileName, $options)
...
+    {
+        $validMimetype = false;
+        foreach ($this->previewMimeTypes as $type) {
+            if (fnmatch($type, $mimeType)) {
+                $validMimetype = true;
+                break;
+            }
+        }
+
+        return $validMimetype;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php]: You can omit the `$validMimetype` variable:

```php
foreach ($this->previewMimTypes as $type) {
    if (fnmatch($tye, $mimeTye)) {
        return true;
    }
}

return false;
```

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 8]
```diff
@@ -134,6 +134,9 @@ class MediaManager implements MediaManagerInterface
      */
     private $downloadPath;
 
+    /** @var FFProbe */
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: Split over three lines

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    ) {
+        $this->ffmpeg = $ffmpeg;
+    }
+
+    /**
+     * @param string $file
+     * @param string $time
+     * @param string $destination
+     *
+     * @return bool
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php]: Please add some more descrition in the PHPDoc

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Video;
+
+use FFMpeg\Coordinate\TimeCode;
+use FFMpeg\FFMpeg;
+
+class VideoThumbnailService implements VideoThumbnailServiceInterface
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php]: Add missing PHPDoc

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php (hunk 9 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php]: Actually, if this method is already described in the interface, use `{@inheritdoc}` instead, and describe the method in more detail in the interface

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        $destination = $this->normalizeFilename($destination);
+
+        try {
+            $video = $this->ffmpeg->open($file);
+
+            $timecode = TimeCode::fromString($time);
+
+            $frame = $video->frame($timecode);
+            $frame->save($destination);
+        } catch (\Exception $e) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php]: Isn't there a more specific exception you could catch? And add at least a comment, why you are catching it, but you do nothing.

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+        return file_exists($destination);
+    }
+
+    /**
+     * @param string $video
+     * @param array $times
+     * @param string $destinationPath
+     *
+     * @return array|bool
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailService.php]: Same as above, use `{@inheritdoc}` and describe the method in more detail in the interface.

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+interface VideoThumbnailServiceInterface
+{
+    /**
+     * @param string $file
+     * @param string $time
+     * @param string $destination
+     *
+     * @return bool
+     */
+    public function generate($file, $time, $destination);
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php]: Describe in PHPDoc in more detail

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * @param string $video
+     * @param array $times
+     * @param string $destinationPath
+     *
+     * @return array|bool
+     */
+    public function batchGenerate(
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php]: Same as above

File: src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\MediaBundle\Media\Video;
+
+interface VideoThumbnailServiceInterface
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/Video/VideoThumbnailServiceInterface.php]: Add PHPdoc

File: src/Sulu/Bundle/MediaBundle/Tests/Unit/Video/VideoThumbnailServiceTest.php [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    /**
+     * @var FFMpeg
+     */
+    protected $ffmpeg;
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $this->basePath = __DIR__ . '/../../../bin/downloads/';
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Tests/Unit/Video/VideoThumbnailServiceTest.php]: Do you still need this line?

File: .gitignore [hunk 17]
```diff
@@ -25,6 +25,9 @@ phpunit.xml
 /bin/jackrabbit-standalone*
 /bin/jackrabbit/*
 
+# Downloads
+/src/Sulu/Bundle/MediaBundle/bin/downloads/*
```
Comment (danrot) [inline .gitignore]: I think this is also not necessary anymore

File: src/Sulu/Bundle/MediaBundle/Resources/config/doctrine/FileVersion.orm.xml [hunk 18]
```diff
@@ -21,6 +21,7 @@
         </field>
         <field name="storageOptions" type="text" column="storageOptions" nullable="true" />
         <field name="mimeType" type="string" column="mimeType" length="255" nullable="true" />
+        <field name="properties" type="string" column="properties" length="1000" nullable="true" />
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/config/doctrine/FileVersion.orm.xml]: 1. You've only added the properties here, shouldn't they be removed from the Media Entity xml configuration as well?
2. Has this change be coordinated with @alexander-schranz ?
3. Why is this necessary?

File: src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php [hunk 19]
```diff
@@ -92,11 +93,15 @@ class FormatManager implements FormatManagerInterface
      */
     private $formats;
 
+    /** @var VideoThumbnailServiceInterface */
+    private $videoThumbnailService;
+
     /**
      * @param MediaRepository         $mediaRepository
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Media/FormatManager/FormatManager.php]: I know it is not your fault, but still: can you reformat this file, so that the alignment in this PHPDoc is gone?

File: src/Sulu/Bundle/MediaBundle/Resources/config/doctrine/FileVersion.orm.xml (hunk 18 above)
Comment (stemod) [inline src/Sulu/Bundle/MediaBundle/Resources/config/doctrine/FileVersion.orm.xml]: 1. Media Entity does not have 'properties' in XML.
2. Yes
3. Videos: saving duration;  could also be used for e.g. PDF -> amount of pages, etc.
